add ram observations

christopherhesse opened the pull request:
Add RAM observations.

It looks like `reset()` used `self.get_screen()` instead of `self.em.get_screen()`, the crop is not supposed to apply to observations, right @endrift?

endrift commented on README.md:
```diff
@@ -9,7 +9,7 @@ Supported platforms:
 
 - Windows 7, 8, 10
 - macOS 10.12 (Sierra), 10.13 (High Sierra), 10.14 (Mojave)
-- Linux
```
Should be a separate commit

christopherhesse replied:
Removed

endrift commented on examples/random_agent.py:
```diff
@@ -4,16 +4,18 @@
 
 parser = argparse.ArgumentParser()
-parser.add_argument('game', help='the name or path for the game to run')
-parser.add_argument('state', nargs='?', help='the initial state file to load, minus the extension')
```
Why these changes? Definitely should be a separate commit.

christopherhesse replied:
Removed

endrift commented on retro/retro_env.py:
```diff
@@ -191,8 +206,10 @@ def step(self, a):
             self.movie.step()
         self.em.step()
-        self.img = ob = self.get_screen()
+        self.img = self.em.get_screen()
```
Crops are intentional.

christopherhesse replied:
My bad, I thought it was inconsistent about crops, but I must have mis-read the code.

endrift commented on retro/retro_env.py:
```diff
@@ -208,9 +225,11 @@ def reset(self):
         if self.movie:
             self.movie.step()
-        self.img = ob = self.get_screen()
+        self.img = self.em.get_screen()
```
Ditto.

christopherhesse replied:
Fixed

christopherhesse commented:
I fixed the comments but also refactored `_get_obs` into `_update_obs` since this means I only need to get the requested type of observation and also `_get_obs` seemed to be only called right after updating.

endrift commented on retro/retro_env.py:
```diff
@@ -191,8 +208,8 @@ def step(self, a):
         if self.movie:
             self.movie.step()
         self.em.step()
-        self.img = ob = self.get_screen()
```
`self.img` still needs to be updated for `env.render()` to work, further the check at the beginning of `env.step` will need to be updated.

christopherhesse replied:
Good call, thanks for the review!

JoeOsborn commented:
This is exciting!  I have some use cases that need both RAM and the framebuffer—would it be possible to get, say, the framebuffer appended onto the RAM snapshot, or a 2-tuple of the two types of data?

endrift commented:
Yeah, it may make sense to add a `BOTH` type that gives a DictSpace.

christopherhesse commented:
A Dict space makes the most sense, but I think we can save that for another issue.   The user should be able to call `env.get_ram()` whenever so that could be called manually or as part of a wrapper that exposes both the image and the ram.

christopherhesse commented:
@endrift it seems I am not added to the travis-ci.org retro project, could you add me please?

christopherhesse commented:
Seems to work now, thanks!
